processor/spanmetricsprocessor/processor.go

```diff
@@ -141,15 +198,196 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+		dpCalls := pdata.NewIntDataPoint()
+		dpCalls.SetStartTime(pdata.TimeToUnixNano(p.startTime))
+		dpCalls.SetTimestamp(pdata.TimeToUnixNano(time.Now()))
+		dpCalls.SetValue(p.callSum[key])
+
+		dpCalls.LabelsMap().InitFromMap(p.metricKeyToDimensions[key])
+
+		mCalls := pdata.NewMetric()
+		mCalls.SetDataType(pdata.MetricDataTypeIntSum)
+		mCalls.SetName("calls")
+		mCalls.IntSum().DataPoints().Append(dpCalls)
+		mCalls.IntSum().SetAggregationTemporality(pdata.AggregationTemporalityCumulative)
+		ilm.Metrics().Append(mCalls)
+	}
+}
+
 // aggregateMetrics aggregates the raw metrics from the input trace data.
 // Each metric is identified by a key that is built from the service name
 // and span metadata such as operation, kind, status_code and any additional
 // dimensions the user has configured.
 func (p *processorImp) aggregateMetrics(traces pdata.Traces) {
-	// TODO: Add implementation
+	for i := 0; i < traces.ResourceSpans().Len(); i++ {
+		rspans := traces.ResourceSpans().At(i)
+		r := rspans.Resource()
+
+		attr, ok := r.Attributes().Get(conventions.AttributeServiceName)
+		if !ok {
+			continue
+		}
+		serviceName := attr.StringVal()
+		p.aggregateMetricsForServiceSpans(rspans, serviceName)
+	}
+}
+
+func (p *processorImp) aggregateMetricsForServiceSpans(rspans pdata.ResourceSpans, serviceName string) {
+	ilsSlice := rspans.InstrumentationLibrarySpans()
+	for j := 0; j < ilsSlice.Len(); j++ {
+		ils := ilsSlice.At(j)
+		spans := ils.Spans()
+		for k := 0; k < spans.Len(); k++ {
+			span := spans.At(k)
+			p.aggregateMetricsForSpan(serviceName, span)
+		}
+	}
+}
+
+func (p *processorImp) aggregateMetricsForSpan(serviceName string, span pdata.Span) {
+	latencyInMilliseconds := float64(span.EndTime()-span.StartTime()) / float64(time.Millisecond.Nanoseconds())
+
+	// Binary search to find the latencyInMilliseconds bucket index.
+	index := sort.SearchFloat64s(p.latencyBounds, latencyInMilliseconds)
+
+	p.lock.Lock()
+	key := p.buildKey(serviceName, span)
+	p.updateCallMetrics(key)
+	p.updateLatencyMetrics(key, latencyInMilliseconds, index)
+	p.lock.Unlock()
+}
+
+// updateCallMetrics increments the call count for the given metric key.
+func (p *processorImp) updateCallMetrics(key string) {
+	p.callSum[key]++
+}
+
+// updateLatencyMetrics increments the histogram counts for the given metric key and bucket index.
+func (p *processorImp) updateLatencyMetrics(key string, latency float64, index int) {
+	if _, ok := p.latencyBucketCounts[key]; !ok {
+		p.latencyBucketCounts[key] = make([]uint64, len(p.latencyBounds))
+	}
+	p.latencySum[key] += latency
+	p.latencyCount[key]++
+	p.latencyBucketCounts[key][index]++
+}
+
+func (p *processorImp) addDimension(key, value string) {
+	// It's worth noting that from pprof benchmarks, WriteString is the most expensive operation of this processor.
+	// Specifically, the need to grow the underlying []byte slice to make room for the appended string.
+	p.metricKeyBuilder.WriteString(value)
```

Comment:
I think we need to delimit the values otherwise we can have key collisions for different dimensions combinations. For example {serviceName="a", operationKey="bc"} will produce the same concatenated key as {serviceName="ab", operationKey="c"} (both result in concatenation of "abc").
Going with a 0 byte as a character should be good enough. 0 byte should not occur in the string value legitimately.
Alternatively, if we think 0 byte can occur in valid attribute values then we will need to escape the string values to ensure the delimiter we use can not be observed in the escaped value alone.

Reply:
Great catch, @tigrannajaryan. I'll try the 0 byte option.